storage/merkle/proof_test.go

```diff
@@ -0,0 +1,166 @@
+// when trie includes only a single value
+func TestProofWithASingleKey(t *testing.T) {
+
+	keyLength := 32
+	tree1, err := NewTree(keyLength)
+	assert.NoError(t, err)
+
+	key, val := randomKeyValuePair(32, 128)
+
+	replaced, err := tree1.Put(key, val)
+	assert.NoError(t, err)
+	require.False(t, replaced)
+
+	// work for an existing key
+	proof, existed := tree1.Prove(key)
+	require.True(t, existed)
+
+	isValid, err := proof.Verify(tree1.Hash())
+	assert.NoError(t, err)
+	require.True(t, isValid)
+
+	// fail for non-existing key
+	key2, _ := randomKeyValuePair(32, 128)
+
+	proof, existed = tree1.Prove(key2)
+	require.False(t, existed)
+	require.Nil(t, proof)
+}
+
+// TestValididateFormat tests cases a proof can not be valid
+func TestValididateFormat(t *testing.T) {
+
+	// construct a valid proof
+	keyLength := 32
+	key := make([]byte, keyLength)
+	key[0] = uint8(5)
+	value := make([]byte, 128)
+	value[0] = uint8(6)
+
+	key2 := make([]byte, keyLength)
+	key2[0] = uint8(4)
+	value2 := make([]byte, 128)
+
+	tree1, err := NewTree(keyLength)
+	assert.NoError(t, err)
+	replaced, err := tree1.Put(key, value)
+	assert.NoError(t, err)
+	require.False(t, replaced)
+	replaced, err = tree1.Put(key2, value2)
+	assert.NoError(t, err)
+	require.False(t, replaced)
+	proof, existed := tree1.Prove(key)
+	require.True(t, existed)
+
+	// invalid key size
+	proof.Key = make([]byte, 0)
+	err = proof.valididateFormat()
+	assert.Error(t, err)
+	require.Equal(t, err.Error(), "malformed proof, key is empty")
+
+	// invalid key size (too large)
+	proof.Key = make([]byte, maxKeyLength+1)
+	err = proof.valididateFormat()
+	assert.Error(t, err)
+	require.Equal(t, err.Error(), "malformed proof, key length is larger than max key length allowed (8193 > 8192)")
+
+	// issue with the key size not matching the rest of the proof
+	proof.Key = make([]byte, 64)
+	err = proof.valididateFormat()
+	assert.Error(t, err)
+	require.Equal(t, err.Error(), "malformed proof, key length doesn't match the length of ShortPathLengths and SiblingHashes")
+
+	// reset the key back to its original value
+	proof.Key = key
+
+	// empty InterimNodeTypes
+	InterimNodeTypesBackup := proof.InterimNodeTypes
+	proof.InterimNodeTypes = make([]byte, 0)
+	err = proof.valididateFormat()
+	assert.Error(t, err)
+	require.Equal(t, err.Error(), "malformed proof, InterimNodeTypes is empty")
+
+	// empty InterimNodeTypes
+	proof.InterimNodeTypes = make([]byte, maxKeyLength+1)
+	err = proof.valididateFormat()
+	assert.Error(t, err)
+	require.Equal(t, err.Error(), "malformed proof, InterimNodeTypes is larger than max key length allowed (8193 > 8192)")
+
+	// issue with the size of InterimNodeTypes
+	proof.InterimNodeTypes = append(InterimNodeTypesBackup, byte(0))
+	err = proof.valididateFormat()
+	assert.Error(t, err)
+	require.Equal(t, err.Error(), "malformed proof, the length of InterimNodeTypes doesn't match the length of ShortPathLengths and SiblingHashes")
+
+	proof.InterimNodeTypes = InterimNodeTypesBackup
```

Comment:
Do we have tests for the following failure conditions (?)
1. auxiliary bits at the end of `InterimNodeTypes` (to pad it to full bytes) not set to zero 
2. number of bits set to `1` in `InterimNodeTypes` different than `ShortPathLengths`
3. number of bits set to `0` in `InterimNodeTypes` (ignoring auxiliary bits at the end) different than `SiblingHashes`
   - comment: we test that a proof is rejected where `InterimNodeTypes` is too long here: https://github.com/onflow/flow-go/blob/d36176260dbe89bce7cb2fe8bf391487feb4044b/storage/merkle/proof_test.go#L102 But I specifically means to test the case, where it has the right size, but the counts of `0`s and `1`s don't match up

Reply:
1. I'll add one 
2. we do actually have it, the error message was not informative, so I updated 
3. I tried to add one but it was always exiting another path (step size check or shortPathLengths), I'll try again